l5kit/l5kit/dataset/ego.py

```diff
@@ -246,3 +253,242 @@ def _get_sample_function(self) -> Callable[..., dict]:
     def get_scene_dataset(self, scene_index: int) -> "EgoDatasetVectorized":
         dataset = self.dataset.get_scene_dataset(scene_index)
         return EgoDatasetVectorized(self.cfg, dataset, self.vectorizer, self.perturbation)
+
+
+class EgoAgentDatasetVectorized(EgoDatasetVectorized):
+    def __init__(
+        self,
+        cfg: dict,
+        zarr_dataset: ChunkedDataset,
+        vectorizer: Vectorizer,
+        perturbation: Optional[Perturbation] = None,
+        agents_mask: Optional[np.ndarray] = None,
+        eval_mode: bool = False
+    ):
+        """
+        Get a PyTorch dataset object that can be used to train DNNs with vectorized input.
+        Ego features are added to the agent features to treat ego as an agent.
+
+        Args:
+            cfg (dict): configuration file
+            zarr_dataset (ChunkedDataset): the raw zarr dataset
+            vectorizer (Vectorizer): a object that supports vectorization around an AV
+            perturbation (Optional[Perturbation]): an object that takes care of applying trajectory perturbations.
+                None if not desired
+            agents_mask (Optional[np.ndarray]): custom boolean mask of the agent availability.
+            eval_mode (bool): enable eval mode (iterates over agent, similarly to AgentDataset).
+        """
+        super().__init__(cfg, zarr_dataset, vectorizer, perturbation)
+
+        self.eval_mode = eval_mode
+        self.vectorizer.other_agents_num -= 1  # account for ego as additional agent
+
+        min_frame_history = AGENT_MIN_FRAME_HISTORY
+        min_frame_future = AGENT_MIN_FRAME_FUTURE
+        if agents_mask is None:  # if not provided try to load it from the zarr
+            agents_mask = self.load_agents_mask()
+            past_mask = agents_mask[:, 0] >= min_frame_history
+            future_mask = agents_mask[:, 1] >= min_frame_future
+            agents_mask = past_mask * future_mask
+
+            if min_frame_history != AGENT_MIN_FRAME_HISTORY:
+                warnings.warn(
+                    f"you're running with custom min_frame_history of {min_frame_history}",
+                    RuntimeWarning,
+                    stacklevel=2,
+                )
+            if min_frame_future != AGENT_MIN_FRAME_FUTURE:
+                warnings.warn(
+                    f"you're running with custom min_frame_future of {min_frame_future}", RuntimeWarning, stacklevel=2
+                )
+        else:
+            warnings.warn("you're running with a custom agents_mask", RuntimeWarning, stacklevel=2)
+
+        # store the valid agents indices (N_valid_agents,)
+        self.agents_indices = np.nonzero(agents_mask)[0]
+
+        # store an array where valid indices have increasing numbers and the rest is -1 (N_total_agents,)
+        self.mask_indices = agents_mask.copy().astype(np.int)
```

Comment:
Set the `copy=True` explicitly in the `astype` and then remove the extra copy `copy()`, by default the `astype` will make a copy, so there is one extra copy here that you can avoid.